include/unifex/detach_on_cancel.hpp

```diff
@@ -0,0 +1,172 @@
+#pragma once
+
+#include <unifex/get_stop_token.hpp>
+#include <unifex/receiver_concepts.hpp>
+#include <unifex/sender_concepts.hpp>
+#include <atomic>
+#include <exception>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+namespace _detach_on_cancel {
+
+template <typename UpstreamSender, typename DownstreamReceiver>
+struct operation_state {
+  struct _receiver;
+
+  struct detached_state;
+
+  struct cancel_callback final {
+    detached_state* state_;
+
+    void operator()() noexcept {
+      auto* op = state_->parentOp_.exchange(nullptr);
+      if (op != nullptr) {
+        // We won the race for cancellation
+        op->state_->stopSource_.request_stop();
```

Comment:
I think this call may need to be performed before the call to `state_->parentOp_->exchange()` above.
As soon as you have exchanged the parentOp with null it's possible that the operation may complete concurrently on another thread, see that parentOp is now null and then destroy the detached_state object (which means this call might be made on a dangling pointer).

If you move it before the call to `exchange()` above then if the operation completes concurrently on another thread then if it sees a null parentOp it will be safe to destroy, or if it sees a non-null parentOp it will then call .destruct() on the callback which will block waiting for the cancel callback to complete (and thus the call to `stopSource_.request_stop()` will have returned and won't be operating on a dangling object).